### ISSUE TITLE
Expression evaluation for objects which can have varying schema on runtime.

### ISSUE THREAD

**Issue by Devanshu24** [cel-expr/cel-go#754]
## Summary
TL; DR: Ability to type check objects that have their fields decided on runtime.
We have a use case to evaluate expressions involving customizable objects. Say you have a base object, `person`.
```
person:
  fields:
    - field_name: name
      field_type: string
    - field_name: age
      field_type: int
```
Now this person object could potentially have custom fields that can only be determined by making an API call when the actual expression is evaluated.
To be specific, you would have to make an API call like `GetCustomPersonFields`, which returns all the custom fields in `Person` along with their types.

## Example
Say we have the expression
`person.name == "John Doe" && person.age == 42`, since this expression only involves the fields in the base Person object, we can always evaluate it, and hence the usual process of creating a checked AST and caching it would work.

Say we have the expression
`person.name == "John Doe" && person.birth_timestamp == "2000-01-01T11:11:11"` this involves a custom field `birth_timestamp` which is of the type `timestamp`.
At environment creation, we do not know about the field's existence or type, so the usual `env.Compile()` would not be helpful as we would not know what to pass to the `cel.Variable`.

## Possible Solutions
One way is to create a new environment each time we need to evaluate the expression but I believe would be very inefficient.

We would want some way in which we can do a precheck for things we know, in the above example that would be the `person.name=="John Doe"` sub-expression and then do some form of an extra check on the partially checked AST when doing the `prg.Eval()`  call.

```golang
env, err := cel.NewEnv(cel.Variable("person", cel.ObjectType("proto_message_for_base_person_type")))

// Only check the types for the base person type in the env.Compile.
ast, iss := env.Compile(`person.name == "John Doe" && person.birth_timestamp == "2000-01-01T11:11:11"`)
prg, err := env.Program(ast)
// Save prg to cache.
err = cache.Set("some_key", prg)

// Load prg from cache.
prg = cache.Get("some_key")
personObject := map[string]interface{}{
	"person": map[string]interface{}{
		"name":            "John Doe",
		"age":             42,
		"birth_timestamp": "2000-01-01T11:11:11",
	},
}

// Validate the prg for the custom fields in the actual person object.
err = prg.Check(personObject, getCustomFieldsClient) // getCustomFieldsClient can be used to invoke GetCustomPersonFields.
out, _, err := prg.Eval(personObject)
fmt.Println(out)
```
Something along these lines.

----

- [x] There are no issues that match the desired change
- [x] The change is large enough it can't be addressed with a simple Pull Request

**Comment by TristonianJones** [comment]
Hi @Devanshu24, I would recommend marking `person` as `dyn` if you don't know the schema in advance:

```go
env, err := cel.NewEnv(cel.Variable("person", cel.DynType))
```

Alternatively, if you could make the `person` an opaque type and not permit access to any fields directly, but rather mediate access to the properties of the person through methods:

```go
personType := cel.OpaqueType("Person")
env, err := cel.NewEnv(
    cel.Types(personType),
    cel.Variable("person", personType),
    cel.Function("name", 
        cel.MemberOverload("person_name", []*cel.Type{personType}, cel.StringType,
             cel.UnaryBinding(func(person ref.Val) ref.Val {
                 return types.String(person.Value().(<typename>).GetName())
             }))),
    )
```

This approach is more involved since it requires you to implement a CEL `ref.Type` similar to what was done for the `types/optional.go`; however, it gives you a lot of flexibility in how you resolve fields with some strong typing for the fields which are consistent.

Your other option is to create a cache of environments per person type definition. This would be up to you to sort out, but mirrors the approaches presented in your question.

Hope that helps,

-Tristan
